base/grisu/grisu.jl

```diff
@@ -118,15 +118,20 @@ function Base.show(io::IO, x::Union{Float64,Float32})
     if get(io, :compact, false)
         _show(io, x, PRECISION, 6, x isa Float64, true)
     else
-        _show(io, x, SHORTEST, 0, true, false)
+        _show(io, x, SHORTEST, 0, get(io, :typeinfo, Any) !== typeof(x), false)
     end
 end
 
 function Base.show(io::IO, x::Float16)
-    if get(io, :compact, false)
+    hastypeinfo = Float16 === get(io, :typeinfo, Any)
+    # if hastypeinfo, the printing will be more compact using `SHORTEST`
+    # BUT: we want to print all digits in `show`, not in display, so we rely
```

Comment:
> All we have to do here is follow what the IOContext properties request

Yes, but we don't have an `IOContext` property to decide how many digits to display, and till now it was conflated with `:compact`. I wanted to change that here, but then Jeff said it's better to show more digits in `show` than in `display` (if I understood correctly what he said in the linked comment in the next lines). So I changed back to the more conservative solution, but noting in this comment why we do so. 

> "if hastypeinfo, the printing will be more compact", since "compact" is a different property. 

It means that if `hastypeinfo`, then the type won't have to be printed again and using `SHORTEST` would allow more "compact" representation in the `:compact` meaning, as in "takes less space on screen". I will try to clarify this comment.

Reply:
Hmm, I'm not sure what Jeff meant exactly, but as I said I'd rather not mention `show` and `display` here. The behavior should be natural just depending on whether `compact` is true or not, independently of what higher-level function called us.

Reply:
> The behavior should be natural just depending on whether compact is true or not

But the problem is that it's not natural, if we are conservative and follow Jeff's suggestion, which I will repeat here:

> I think we should always print all digits, except in a context that's clearly for convenient interactive use (i.e. display).

So the only solution I saw is to make a comment here to explain why we do what we do here, until a new property is introduced which make the current behavior "natural".

Reply:
Yes, but probably better leave this tweaking issue aside for now. There are already enough changes in this PR.

Reply:
I agree, but I thought what I did is precisely that... are you fine with the current state, or are you suggesting a specific change here?

Reply:
I really don't know. Let's leave it that way.